Add Cloud Native AI Scheduling Challenges Whitepaper (#2164)
cncf/toc: initiatives/1641_Cloud_Native_AI_Scheduling_Challenges_Whitepaper/5_Solutions_and_Practical_Guidance_for_AI_Workload_Scheduling.md

* **KAI Scheduler** (formerly Run:ai Scheduler, a CNCF sandbox project) is an AI workload scheduler designed for large scale, high throughput clusters. Features include gang scheduling, topology awareness, hierarchical queues with quotas, fair-share policies (with/without time based), GPU sharing (partitioning, time-slicing), elastic workloads, and multi-framework support. Used in enterprise deployments with thousands of nodes.   
* **Apache YuniKorn** provides unified scheduling for batch and interactive workloads with hierarchical queues and fair sharing.  
* **Slinky** (opensource project maintained by NVIDIA) is a set of powerful integration tools designed to bring Slurm capabilities into Kubernetes, both allowing Slurm jobs to run within Kubernetes and bringing the full suite of Slurm scheduling capabilities into Kubernetes.  
* 

## ML Platform Tools

> These tools provide higher-level abstractions for ML workflows:

Reviewer: ```suggestion
These tools provide higher-level abstractions for AI workloads:
```

* **Kubeflow**  

>   * **Kubeflow Trainer** supports distributed training across frameworks (PyTorch, TensorFlow, PaddlePaddle, XGBoost). Provides job abstractions that handle worker coordination, including gang scheduling requirements.  

Reviewer: Ref: https://github.com/kubeflow/trainer#overview

```suggestion
  * **Kubeflow Trainer** is a Kubernetes-native distributed AI platform for scalable LLM fine-tuning and training of AI models across a wide range of frameworks, including PyTorch, MLX, HuggingFace, DeepSpeed, JAX, XGBoost, and more. Provides job abstractions that handle worker coordination, including gang scheduling requirements and HPC workloads orchestration such as MPI and Flux.
```

# Mapping Challenges to Solutions

A reference table that maps each challenge to the solutions that address it. All solutions in alphabetical order.

| Challenge | Kubernetes Native | Schedulers | Platforms/Tools | Applies To | Notes |
| :---- | :---- | :---- | :---- | :---- | :---- |
| Gang Scheduling | Workload API \+ GangScheduling feature gate (alpha, K8s 1.35+) | Armada, Coscheduling plugin, KAI, Kueue, Slinky, Volcano, YuniKorn | Kubeflow Trainer | Both | Essential for distributed training |
| Resource Fairness | ResourceQuota (limited) | Armada, KAI, Kueue, Slinky, Volcano, YuniKorn | \- | Both | Hierarchical quotas require external tools |
| Queue Management | \- | Armada, KAI, Kueue, Slinky, Volcano, YuniKorn  | Airflow, Flyte | Both | Core capability of batch schedulers |
| Preemption | PriorityClass (pod-level) | KAI, Kueue, Slinky, Volcano | \- | Both | Job-level preemption needs external tools |
| Priority Scheduling | PriorityClass | All batch schedulers | \- | Both | Job-level priority in batch schedulers |
| Reservation & Backfill | \- | Slinky, Volcano, YuniKorn | \- | Training | Advanced feature in some schedulers |

> | Topology Awareness (Node) | Topology Manager (NUMA), DRA CPU Driver (CPU topology) | KAI, Kueue, Slinky, Volcano | \- | Both | GPU interconnect awareness varies |

Reviewer 2: May i suggest to add HAMi here, HAMi also has the node-level topology-awareness(https://project-hami.io/docs/userguide/nvidia-device/scheduling-policy)

> | Preemption | PriorityClass (pod-level) | KAI, Kueue, Slinky, Volcano | \- | Both | Job-level preemption needs external tools |
> | Priority Scheduling | PriorityClass | All batch schedulers | \- | Both | Job-level priority in batch schedulers |
> | Reservation & Backfill | \- | Slinky, Volcano, YuniKorn | \- | Training | Advanced feature in some schedulers |
> | Topology Awareness (Node) | Topology Manager (NUMA), DRA CPU Driver (CPU topology) | KAI, Kueue, Slinky, Volcano | \- | Both | GPU interconnect awareness varies |
> | Topology Awareness (Cluster) | Topology Spread Constraints, DRANET (network DRA Driver) (limited) | KAI, Kueue, Slinky, Volcano | \- | Both | Network topology awareness is emerging |

Reviewer: I think for Preemption and TAS all platforms that integrate with Kueue should support it including KubeRay, Trainer, and k8s
WDYT @rajaskakodkar ?

```suggestion
| Preemption | PriorityClass (pod-level) | KAI, Kueue, Slinky, Volcano | KubeRay, Kubeflow Trainer, Kubernetes | Both | Job-level preemption needs external tools |
| Priority Scheduling | PriorityClass | All batch schedulers | KubeRay, Kubeflow Trainer, Kubernetes | Both | Job-level priority in batch schedulers |
| Reservation & Backfill | \- | Slinky, Volcano, YuniKorn | KubeRay, Kubeflow Trainer, Kubernetes | Training | Advanced feature in some schedulers |
| Topology Awareness (Node) | Topology Manager (NUMA), DRA CPU Driver (CPU topology) | KAI, Kueue, Slinky, Volcano | KubeRay, Kubeflow Trainer, Kubernetes | Both | GPU interconnect awareness varies |
| Topology Awareness (Cluster) | Topology Spread Constraints, DRANET (network DRA Driver) (limited) | KAI, Kueue, Slinky, Volcano | KubeRay, Kubeflow Trainer, Kubernetes | Both | Network topology awareness is emerging |
```

Author: This makes sense. Thanks for the suggestion.

| Resource Heterogeneity | Node selectors, labels | All batch schedulers | \- | Both | Standard Kubernetes features usually sufficient |
| GPU Sharing | DRA (GA, K8s 1.34+) | KAI | HAMi, KubeRay, Volcano | Both | MIG requires DRA or vendor tools |

> | Scalability | Cluster Autoscaler, Karpenter | Armada, KAI, Kueue, Slinky, Volcano | interLink | Both | Large-scale scheduling is challenging |

Reviewer 3: The Scalability row/inference req autoscaling in the solutions table already covers good tools. Should KEDA be included somewhere? Ref: https://github.com/cncf/toc/issues/2188 and https://www.cncf.io/blog/2026/05/27/gpu-autoscaling-on-kubernetes-with-keda-building-an-external-scaler/ . Just a qn coz we started with scheduling, but still the table covers scaling as well.

| I/O Bottlenecks | PersistentVolumes | \- | Fluid | Both | Storage and caching solutions |

> | Fault Tolerance | \- | Slinky,  | Kubeflow (elastic training) | Training | Framework-dependent |

Reviewer: ```suggestion
| Fault Tolerance | \- | Slinky,  | Kubeflow Trainer | Training | Framework-dependent |
```

* Oversubscribing without preemption policies. Promising more resources than exist only works if you have clear rules about who gets preempted when.  
* Ignoring GPU memory constraints. Scheduling multiple workloads to a GPU when they won't fit in memory causes crashes, not sharing.  
* Treating all GPUs as equivalent. Different GPU models, memory sizes, and interconnects matter. Jobs that need current-generation GPUs shouldn't land on previous-generation hardware.   
* No quotas in multi-tenant clusters. Without quotas, one team's large job can starve everyone else indefinitely.

### Getting Started

**For platform engineers building new infrastructure:**

1. Start with a single batch scheduler to handle basic queuing and quota management.  
2. Define queues that map to your organizational structure (teams, projects, environments).  
3. Set quotas based on expected resource allocation—you can adjust later.  
4. Ensure gang scheduling support if you run distributed training.  
5. Implement monitoring to understand actual resource usage patterns.

**For ML engineers working with existing infrastructure:**

1. Understand what scheduling tools are available in your cluster.  

> 2. Use the appropriate job abstractions (PyTorchJob, MPIJob, etc.) rather than raw pods.  

Reviewer: ```suggestion
2. Use the appropriate job abstractions (TrainJob, MPIJob, PyTorchJob, etc.) rather than raw pods.  
```

Reviewer 4: suggest to also keep PyTorchJob for legacy environments

Reviewer: Sounds good. I changed the suggetsion @rajaskakodkar.

1. Understand your workloads. Different stages of the ML lifecycle have different requirements. Data preparation, training, and inference each need different scheduling strategies.  
2. Gang scheduling is essential for AI workloads. Without it, partial allocations waste resources and cause deadlocks. If you run distributed training, you need a scheduler that supports gang scheduling.  
3. Fairness requires explicit policies. ResourceQuota provides basic limits, but real multi-tenant clusters need hierarchical quotas, borrowing/lending, and fair-share algorithms.  
4. Topology matters. For communication-intensive workloads, placing workers on well-connected nodes significantly improves performance. This is an area of active development.  
5. The ecosystem is maturing. Projects like Kueue, Slinky, Volcano, and KAI Scheduler are production-ready. Dynamic Resource Allocation is reaching stability. The tools exist—the question is choosing and integrating them.  
6. Start simple, iterate. You don't need every feature on day one. Start with basic queuing and quotas, and add capabilities as your needs grow.

> The cloud-native AI landscape continues to evolve. New hardware (different GPU architectures, accelerators), new workload patterns (agentic systems, distributed inference), and new Kubernetes features will create both challenges and opportunities. This series provides a foundation; staying current requires ongoing engagement with the community.

Reviewer 4: > This series provides a foundation; staying current requires ongoing engagement with the community.

love this ❤️
